lib/opencensus/trace/span_builder.rb

```diff
@@ -314,52 +348,240 @@ def initialize span_context, sampled, skip_frames: 0
+              OpenCensus::Trace::Annotation.new(
+                truncatable_string(ann.description),
+                attributes: attrs,
+                dropped_attributes_count: dropped_attributes_count,
+                time: ann.time
+              )
+            count += 1
+          end
+          [result, raw_annotations.size - count]
+        end
+
+        ##
+        # Build a canonical message list and the number of dropped items
+        # @private
+        #
+        def message_event_pieces raw_message_events
+          result = []
+          count = 0
+          raw_message_events.each do |evt|
+            break if @max_message_events != 0 && count == @max_message_events
+            result <<
+              OpenCensus::Trace::MessageEvent.new(
+                evt.type,
+                evt.id,
+                evt.uncompressed_size,
+                compressed_size: evt.compressed_size,
+                time: evt.time
+              )
+            count += 1
+          end
+          [result, raw_message_events.size - count]
         end
-      end
 
-      def status
-        return nil unless @status_code || @status_message
+        ##
+        # Build a canonical links list and the number of dropped items
+        # @private
+        #
+        def link_pieces raw_links
+          result = []
+          count = 0
+          raw_links.each do |lnk|
+            break if @max_links != 0 && count == @max_links
+            attrs, dropped_attributes_count = attribute_pieces lnk.attributes
+            result <<
+              OpenCensus::Trace::Link.new(
+                lnk.trace_id,
+                lnk.span_id,
+                type: lnk.type,
+                attributes: attrs,
+                dropped_attributes_count: dropped_attributes_count
+              )
+            count += 1
+          end
+          [result, raw_links.size - count]
+        end
 
-        Status.new @status_code, @status_message
+        ##
+        # Build a canonical status object
+        # @private
+        #
+        def status_piece status_code, status_message
+          return nil unless status_code || status_message
+          Status.new status_code.to_i, status_message.to_s
+        end
+
+        ##
+        # Build a truncatable string
+        # @private
+        #
+        def truncatable_string str
+          return str if str.is_a? TruncatableString
+          orig_str = str.encode Encoding::UTF_8,
+                                invalid: :replace,
+                                undef: :replace
+          if @max_string_length != 0 && @max_string_length < str.bytesize
+            str = truncate_str orig_str, @max_string_length
+            truncated_bytes = orig_str.bytesize - str.bytesize
+            TruncatableString.new str, truncated_byte_count: truncated_bytes
+          else
+            TruncatableString.new orig_str
+          end
+        end
+
+        private
+
+        def truncate_str str, target_bytes
```

Comment:
Have you seen how Rails handles this? Basically, split into bytes, truncate at the desired byte length, then unpack as unicode. If it throws a `ArgumentError` (ArgumentError: malformed UTF-8 character), try at desired byte length - 1, until it succeeds unpacking.

This implementation seems like it's doing extra work and is conceptually harder to grasp.

Reply:
Good call. Done.